Мордвинцева Алена StarWars

mordvintseva opened the pull request.

Commit:
First homework. Initial commit

M0rtyMerr commented on Mordvintseva/StarWars/First_app/CharacterViewController.swift:
```diff
@@ -0,0 +1,68 @@
+class CharacterViewController: UIViewController, UITableViewDataSource {
+    @IBOutlet private var nameLabel: UILabel!
+    @IBOutlet private var personalFeaturesTable: UITableView!
+    let personalFeatureCellIdentifier = "personalFeature"
```
Тоже Private, другим классам не нужно знать идентификтор ячейки

M0rtyMerr commented on Mordvintseva/StarWars/First_app/CharacterViewController.swift:
```diff
@@ -0,0 +1,68 @@
+    let personalFeatureCellIdentifier = "personalFeature"
+    private var character = Character()
+
+    enum PersonalFeatures: Int, CaseIterable {
```
Уважаю но private

M0rtyMerr commented on Mordvintseva/StarWars/First_app/CharacterViewController.swift:
```diff
@@ -0,0 +1,68 @@
+
+    enum PersonalFeatures: Int, CaseIterable {
+        case height = 0
+        case mass, hair_color, skin_color, eye_color, birth_year, gender
```
Мяяя, не очень. Принято camel case, то есть hairColor

M0rtyMerr commented on Mordvintseva/StarWars/First_app/CharacterViewController.swift:
```diff
@@ -0,0 +1,68 @@
+        self.nameLabel.text = character.name
+        self.personalFeaturesTable.dataSource = self
+        DispatchQueue.main.async {
+            self.personalFeaturesTable.reloadData()
```
В данном случае это DispatchQueue.main.async не нужен. viewDidLoad() вызывается на главном потоке всегда
В моем примере мы дергали асинхронный метод, который вызывался не на главном потоке

M0rtyMerr commented on Mordvintseva/StarWars/First_app/CharacterViewController.swift:
```diff
@@ -0,0 +1,68 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = personalFeaturesTable.dequeueReusableCell(withIdentifier: personalFeatureCellIdentifier, for: indexPath)
+        let num: PersonalFeatures? = PersonalFeatures(rawValue: indexPath.row)
+        var category = ""
```
Идея понятна, но можно лучше. Зачем тебе передавать пустой текст в ячейку?
Напиши просто `guard let num = PersonalFeatures(rawValue: indexPath.row) else { return cell)`

M0rtyMerr commented on Mordvintseva/StarWars/First_app/CharacterViewController.swift:
```diff
@@ -0,0 +1,68 @@
+        let cell = personalFeaturesTable.dequeueReusableCell(withIdentifier: personalFeatureCellIdentifier, for: indexPath)
+        let num: PersonalFeatures? = PersonalFeatures(rawValue: indexPath.row)
+        var category = ""
+        var value = ""
```
конструкции вида
```
let value: String
if someBool {
  value = "1"
} else {
  value - "2"
}
```
компилятся (и со switch тоже). Они лучше, потому что переменная иммутабельна

M0rtyMerr commented on Mordvintseva/StarWars/First_app/Model/Character.swift:
```diff
@@ -0,0 +1,47 @@
+    let url: String
+
+    init() {
+        name = ""
```
Не очень понял зачем это?

mordvintseva replied:
Где-то создавала переменную, в completionHandler её обрабатывала, а вне его выводила. На меня, вроде, ругался Xcode, потому что она не инициализирована, хотя я перепроверю

M0rtyMerr commented on Mordvintseva/StarWars/First_app/Model/Item.swift:
```diff
@@ -0,0 +1,9 @@
+//  Copyright © 2019 Mordvintseva Alena. All rights reserved.
+//
+
+import Foundation
```
Лишние файлы удаляем :)

M0rtyMerr commented on Mordvintseva/StarWars/First_app/Service/CharacterServiceNetwork.swift:
```diff
@@ -0,0 +1,45 @@
+class CharacterServiceNetwork: CharacterService {
+    func getCharacters(_ completionHandler: @escaping (([Character]) -> Void)) {
+        var characters: [Character] = []
+        var urlAdress: String? = "https://swapi.co/api/people/"
```
А зачем optional? Это совершенно точно валидная строка :)

mordvintseva replied:
По адресу /api/people мне выдают страницу с персонажами и ссылкой на следующую страницу, она может оказаться nil

M0rtyMerr commented on Mordvintseva/StarWars/First_app/Service/CharacterServiceNetwork.swift:
```diff
@@ -0,0 +1,45 @@
+        var characters: [Character] = []
+        var urlAdress: String? = "https://swapi.co/api/people/"
+
+        while let urlString = urlAdress {
```
Странная конструкция, optional не нужен

M0rtyMerr commented on Mordvintseva/StarWars/First_app/Service/CharacterServiceNetwork.swift:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    func getItemByURL<T: Decodable>(url: URL, completionHandler: @escaping ((T) -> Void)) {
+        let semaphore = DispatchSemaphore(value: 0)
```
Какая мотивация у синхронных вызовов?

mordvintseva replied:
Ой, осталось с того момента, как запрос выполнялся в цикле выше, здесь уже не нужно..) Без синхронизации я не могла нормально добавлять персонажей в список (возможно, причина в кривых руках)

M0rtyMerr replied:
По факту ты сейчас блокируешь главный поток. То бишь все твое приложение ждет, пока персонажи скачаются. 
А, так я наконец понял зачем тебе while. Ты хочешь выкачать всех людей за раз в цикле. Давай подумаем, нужно ли это? Если пользователь заходит на главный экран и НЕ скролит таблицу вниз, нужен ли ему 84-ый и 123-ый персонажи?

mordvintseva replied:
В целом, нет, поняла, к чему ты ведёшь)

M0rtyMerr replied:
Подобную технику можно использовать и я даже видел. 
Но тогда уж заворачивай вызов асинхронного запроса в 
```
DispatchQueue.global().async {
    async_blocking_request
}
```
чтобы семафор не блокировал mainThread
Ну и а вообще, на данном этапе развития - если можно сделать без многопоточности, лучше делать без нее. Я не буду тестить корнер кейсы если юзер за сотую секунды пролистает всю таблицу и у тебя несколько параллельных запросов будет

M0rtyMerr commented on Mordvintseva/StarWars/First_app/Service/CharacterServiceNetwork.swift:
```diff
@@ -0,0 +1,45 @@
+        let semaphore = DispatchSemaphore(value: 0)
+        let task = URLSession.shared.dataTask(with: url) { data, _, error in
+            defer { semaphore.signal() }
+            guard let data = data else { return }
```
можно через запятую проверить оба условия

M0rtyMerr commented on Mordvintseva/StarWars/First_app/ViewController.swift:
```diff
@@ -0,0 +1,47 @@
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        // Do any additional setup after loading the view, typically from a nib.
```
Комменты лучше удалять, они не несут никакого смысла

MortyMerr commented:
Приложение работает, минимальные задачи выполнены, за UINavigationController и CaseIterable отдельный лайк
С семафорами интересно. Я не критикую, но мне интересно зачем. Расскажешь?

Commit:
Add .gitignore

Commit:
Code refactoring

Commit:
Change of characters loading

Character loading when reaching the end of UITableView
Delete initializer of Character struct

M0rtyMerr commented on Mordvintseva/StarWars/First_app/CharacterViewController.swift:
```diff
@@ -0,0 +1,68 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = personalFeaturesTable.dequeueReusableCell(withIdentifier: personalFeatureCellIdentifier, for: indexPath)
+        guard let num = PersonalFeatures(rawValue: indexPath.row) else { return cell }
+        let category: String
```
А зачем category здесь, если можно его там ниже сразу и объявить?

MortyMerr commented:
Оценка пять

Commit:
Delete variable category from CharacterViewController
